### ISSUE TITLE
Feature Request: Code Splitting Analyzer

### ISSUE THREAD

**Issue by tihuan** [yannickcr/eslint-plugin-react#2577]
Hi all!

Just wondering if this is the right place to ask if people see the need for a linting rule to detect code splitting opportunity.

For example, the following code shows that `OptionA` and `OptionB` are conditionally needed, so that's an opportunity for code splitting.

```tsx
import OptionA from './OptionA'
import OptionB from './OptionB'

function Component() {
  return option === 'a' ? <OptionA /> : <OptionB />
}
```

Is there a similar tool available already? If not, I'm interested in developing one!

Thank you!

**Comment by ljharb** [comment]
Can you elaborate on what you mean by "code splitting"?

In this example, the advantages of not loading both components up front have to be weighed against the costs of making another http request, and the UX cost of the delay that's visible to the user. It does not strike me as a decision that should be encouraged by tooling.

**Comment by tihuan** [comment]
Hey @ljharb ! Thanks so much for your quick feedback! 

Ah yes, I think the intention is to flag possible code splitting (dynamic import) opportunities at develop time.

But like you said, the engineers will still need to decide if it's the right decision to code split based on the criteria you listed. And if it's not, then a eslint-ignore comment can be added to let other engineers know it's an explicit decision to not code split.

Not sure if the linter is the right place for this, just wanted to reuse the AST and also provide feedback to engineers in real time!

**Comment by ljharb** [comment]
I think you could make a custom rule, but I'm very skeptical of its utility, especially since the majority of cases it identifies (in my experience) would not be good places to do code splitting.

**Comment by tihuan** [comment]
That's a really good point! Sounds like the rule will need to be smarter, otherwise it will just introduce too much false positive and overwhelm people 🤔 

I've never built a linter rule before, how difficult is it to build a prototype to test/tweak this?

I'm thinking we'll need to be able to identify import statement, dynamic import, ternary, and if block to start with?

I guess the pain point I'm trying to address is engineers importing functions and components that are gated behind feature flags or infrequent behaviors (say, only render a calculator when a user clicks on the show calculator button, but only 20% users do so)

So right now it's through training and code reviews to help prevent the bundle bloat. But I wanted to see if it's possible to make the process easier through tools.

Or what other ways have you found effective in addressing the problems?

Thanks again for the awesome feedback!

**Comment by ljharb** [comment]
I’ve not found anything other than code review as an effective tool here - it’s also notable that most apps don’t need any form of code splitting; it’s only very large ones where it starts to be helpful, and you can go a very long way with a small number of manual splits.

**Comment by tihuan** [comment]
Right! We're splitting the app at least by routes and feature flags now, although code review has missed a few ones that we only caught later

I think I'll try to find some time to build a prototype for this custom rule to test it next week and report back how it goes!

Thanks again for your time, Jordan 🤩🙏

**Comment by tihuan** [comment]
Hey @ljharb ! Sorry just circling back to this. I haven't had time to actively look into this given the workload changes and priority shifting from COVID-19 🙏

I'm happy to reopen a new issue or just follow up on this thread later. But for issue hygiene, should we close this issue for now?

Thank you!

**Comment by ljharb** [comment]
I think it's fine to leave it open until you have time to follow up on it; issue hygiene isn't about open issue count, it's about making sure issue statuses are accurate :-)

**Comment by tihuan** [comment]
Haha okay that makes sense 🤩Thanks again and stay healthy!

**Comment by ljharb** [comment]
Closing for now :-) happy to reopen if there's something to do here.
